addons/event_crm/models/event_lead_rule.py

```diff
@@ -0,0 +1,97 @@
+            'One per Order : A single Lead is created per Ticket Batch/Sale Order (B2B)')
+    event_type_ids = fields.Many2many('event.type',
+        string='Event Categories',
+        help='Filter the attendees to include those of this specific event category. If not set, no event category restriction will be applied.')
+    event_id = fields.Many2one('event.event',
+        string='Event', domain="[('company_id', 'in', [company_id or current_company_id, False])]",
+        help='Filter the attendees to include those of this specific event. If not set, no event restriction will be applied.')
+    company_id = fields.Many2one('res.company', string='Company',
+        help="Restrict the trigger of this rule to events belonging to a specific company.\nIf not set, no company restriction will be applied.")
+    event_registration_filter = fields.Text(string="Registrations Domain",
+        help="Filter the attendees that will or not generate leads.")
+
+    # Lead default_value fields
+    lead_type = fields.Selection([
+        ('lead', 'Lead'), ('opportunity', 'Opportunity')], string="Lead Type", required=True,
+        default=lambda self: 'lead' if self.env['res.users'].has_group('crm.group_use_lead') else 'opportunity',
+        help="Default lead type when this rule is applied.")
+    lead_sales_team_id = fields.Many2one('crm.team', string='Sales Team', help="Automatically assign the created leads to this Sales Team.")
+    lead_user_id = fields.Many2one('res.users', string='Salesperson', help="Automatically assign the created leads to this Salesperson.")
+    lead_tag_ids = fields.Many2many('crm.tag', string='Tags', help="Automatically add these tags to the created leads.")
+
+    def _create_leads(self, registrations):
+        """
+        Create leads based on lead creation type of the rule and apply to this leads
+        some prefilled values based on the rule and the registration.
+        """
+        leads = self.env['crm.lead']
+        for rule in self:
+            vals_list = []
+            filtered_registrations = rule._filter_registrations(registrations)
+            if filtered_registrations:
+                if rule.lead_creation_basis == 'attendee':
+                    for registration in filtered_registrations:
+                        vals_list.append(registration._get_registration_lead_values(rule))
+                else:
+                    if not filtered_registrations[0].main_registration_id:
+                        vals_list.append(filtered_registrations._get_registration_lead_values(rule))
+                    else:
+                        # Update the description of the lead linked to the main registration instead of creating a new one
+                        for registration in filtered_registrations:
+                            leads = registration.main_registration_id.lead_ids.filtered(lambda lead: lead.registration_rule_id == rule)
+                            for lead in leads:
```

Comment:
This kind of double for loop makes me think that all informations is duplicated everywhere. Should really be cleaned, because I don't understand why in a loop that creates leads somehow other leads are updated in double for loops.

I think you should try to clean the organization of the method
 * per attendee -> ok
 * per order -> classify registrations per group, create the lead of the master with description based on all its children